### PR TITLE
Fix/ignore typing errors and add mypy to ci

### PR DESCRIPTION
**Pull request by paiforsyth**
Thanks for this package! It looks like it could very useful!

I thought I would contribute by adding type checking.

- Fix some typing errors
- Ignore some other typing errors that seem to be caused by an incompatibility between the use of properties in this project and what mypy expects.
- Add mypy to dev requirements
- Add mypy to CI  using invoke
- Add invoke to dev requirements

I tried to add a mypy pre-commit hook, but this caused pre-commit CI to fail.  I think this may be because the pre-commit CI is not able to install the correct requirements.

### CODE REVIEW

File: xarray_schema/dataarray.py [hunk 1]
```diff
@@ -60,16 +62,16 @@ def __init__(
...
-        self.dtype = dtype
-        self.shape = shape
-        self.dims = dims
-        self.name = name
-        self.coords = coords
-        self.chunks = chunks
-        self.attrs = attrs
-        self.array_type = array_type
-        self.checks = checks
+        # see https://github.com/python/mypy/issues/3004
```
Comment (paiforsyth) [inline xarray_schema/dataarray.py:65]: mypy seems to fail when it is applied to a property whose setter has a different return type than the parameter type of the corresponding getter.

File: xarray_schema/dataset.py [hunk 2]
```diff
@@ -27,7 +28,7 @@ def __init__(
 
         self.data_vars = data_vars
         self.coords = coords
-        self.attrs = attrs
+        self.attrs = AttrSchema(attrs)
```
Comment (jhamman) [inline xarray_schema/dataset.py:31]: what's the reason for needing to do this here, as opposed to in the setter?

File: tasks.py [hunk 3]
```diff
@@ -0,0 +1,9 @@
+from invoke import task
```
Comment (jhamman) [inline tasks.py:1]: I haven't used `invoke` before. What does this do that calling mypy in as a shell command inside GitHub actions does not?

**Review by jhamman — COMMENTED** [review]
@paiforsyth - thank you for the contribution. This all looks great. Two small comments but otherwise all seems good.

File: xarray_schema/dataset.py (hunk 2 above)
Comment (paiforsyth) [reply xarray_schema/dataset.py]: I've fixed this to use a setter

Comment (paiforsyth) [reply xarray_schema/dataset.py]: By the way, have you considered using dataclasses https://docs.python.org/3/library/dataclasses.html or attrs classes https://www.attrs.org/en/stable/ to represent schemas?  They seem a natural fit.

File: tasks.py (hunk 3 above)
Comment (paiforsyth) [reply tasks.py:1]: Invoke's tasks.py acts like a makefile, allowing you to build a collection of standardised commands you can call both in or outside of CI.

Comment (paiforsyth) [reply tasks.py:1]: It is not strictly essential here, but I find it useful.

Comment (paiforsyth) [reply tasks.py:1]: Would you like me to remove it?

Comment (jhamman) [reply tasks.py:1]: Let's give it a try. Seems like a light addition for now. Easy to remove down the line if it becomes a bother.

File: xarray_schema/dataset.py (hunk 2 above)
Comment (jhamman) [reply xarray_schema/dataset.py]: 100% agree that we could transition to datacalsses. I may have even tried this in an earlier version. If memory serves, I ran into a problem with inheritance and decided to roll my own for the purposes of the MVP.

Comment (jhamman) [reply xarray_schema/dataset.py]: If you'd like, we can discuss this idea in separate issue.

**Comment by jhamman** [comment]
closing and reopening to re-trigger codecov

**Comment by paiforsyth** [comment]
I had to mark task.py as ignored by coverage to pass the coverage check

**Comment by jhamman** [comment]
Thanks @paiforsyth. I'm going to merge this now. Cheers!

**Pull request merged**